packages/webchannel-wrapper/gulpfile.js

```diff
@@ -0,0 +1,136 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const gulp = require('gulp');
+const rollup = require('rollup');
+const closureCompiler = require('google-closure-compiler').gulp();
+const del = require('del');
+const path = require('path');
+const sourcemaps = require('gulp-sourcemaps');
+const { resolve } = require('path');
+const commonjs = require('rollup-plugin-commonjs');
+const rollupSourcemaps = require('rollup-plugin-sourcemaps');
+
+// The optimization level for the JS compiler.
+// Valid levels: WHITESPACE_ONLY, SIMPLE_OPTIMIZATIONS, ADVANCED_OPTIMIZATIONS.
+// TODO: Add ability to pass this in as a flag.
+const OPTIMIZATION_LEVEL = 'ADVANCED_OPTIMIZATIONS';
+
+// For minified builds, wrap the output so we avoid leaking global variables.
+// For minified builds, wrap the output so we avoid leaking global variables.
+const CJS_WRAPPER_PREFIX = `(function() {var firebase = require('@firebase/app').default;`;
+const CJS_WRAPPER_SUFFIX =
+  `}).apply(typeof global !== 'undefined' ? ` +
+  `global : typeof self !== 'undefined' ? ` +
+  `self : typeof window !== 'undefined' ? window : {});`;
+
+const closureLibRoot = path.dirname(
+  require.resolve('google-closure-library/package.json')
+);
+
+const closureDefines = [
+  // Avoid unsafe eval() calls (https://github.com/firebase/firebase-js-sdk/issues/798)
+  'goog.json.USE_NATIVE_JSON=true',
+  // Disable debug logging (saves 8780 bytes).
+  'goog.DEBUG=false',
+  // Disable fallbacks for running async code (saves 1472 bytes).
+  'goog.ASSUME_NATIVE_PROMISE=true',
+  // Disables IE8-specific event fallback code (saves 523 bytes).
+  'goog.events.CAPTURE_SIMULATION_MODE=0',
+  // Disable IE-Specific ActiveX fallback for XHRs (saves 524 bytes).
+  'goog.net.XmlHttpDefines.ASSUME_NATIVE_XHR=true'
+];
+
+/**
+ * Builds the core Firebase-auth JS.
+ * @param {string} filename name of the generated file
+ * @param {string} prefix prefix to the compiled code
+ * @param {string} suffix suffix to the compiled code
+ */
+function createBuildTask(filename, prefix, suffix) {
+  return function closureBuild() {
+    return gulp
+      .src(
+        [
+          `${closureLibRoot}/closure/goog/**/*.js`,
+          `${closureLibRoot}/third_party/closure/goog/**/*.js`,
+          'src/**/*.js'
+        ],
+        { base: '.' }
+      )
+      .pipe(sourcemaps.init())
+      .pipe(
+        closureCompiler({
+          js_output_file: filename,
+          output_wrapper: `${prefix}%output%${suffix}`,
+          entry_point: 'firebase.webchannel.wrapper',
+          compilation_level: OPTIMIZATION_LEVEL,
+          externs: [
+            resolve(__dirname, './externs/overrides.js'),
+            resolve(__dirname, './externs/module.js')
+          ],
+          language_out: 'ES5',
```

Comment:
Could this be an option? I wonder if we would get size savings if we used ES2017 for our ES2017 build.

Reply:
Added an ES2017 build, size comparison:

```
47365 packages/webchannel-wrapper/dist/index.esm.js
41267 packages/webchannel-wrapper/dist/index.esm2017.js
47021 packages/webchannel-wrapper/dist/index.js
```

Reply:
Thanks! I think that is worth the effort. Do you know why the SDK size went up, despite this change?

Reply:
The problem was that previously all builds were actually ES2017, due to a bug in `closure-builder`.  This was great for size but not great for ES5 compatibility. Now the builds that are supposed to be ES5 actually are (making them bigger) and a correctly-labeled ES2017 build has been added.

Reply:
Makes sens.

Reply:
I actually thought more about this - the size increase for the NPM builds are explained by these changes. The size increase for the CDN build is not though. The CDN build should have post processed everything to ES5 already. Hence, I think what is going on here is that the new build pipeline is no longer tree shaking code that Firestore isn’t using. Is that a possibility?

Reply:
So my explanation is incorrect for the ESM bundle which was already writing ES5, I think because of using an outdated library (see below).  The bundle size increase is actually due to a change in `closure-library` from versions 20200112 to 20200204. I don't know exactly which change yet, but this led to a 5K bundle size spike on the webchannel-wrapper ES5 builds between those two `closure-library` versions.

Because webchannel-wrapper was using `closure-builder`, the ESM build was fixed on a pretty old version bundled with `closure-builder`.  Once we upgraded it to use a newer version of google-closure-library (20200224, I copied the version auth was using) it saw a 5K size increase.

If we upgrade to the latest version of google-closure-library (20200406) the size increase is smaller (2.5K) but still there.

For now we might want to pin closure-library to 20200112 (before the change that caused this size increase) and try to figure out exactly where the increase is coming from.